nbgrader/apps/zipcollectapp.py

```diff
@@ -0,0 +1,522 @@
+
+        To change the default plugin, you will need a class that inherits from
+        :class:`~nbgrader.plugins.zipcollect.FileNameProcessor`. If your
+        exporter is named `MyCustomProcessor` and is saved in the file
+        `myprocessor.py`, then:
+
+            nbgrader zip_collect --processor=myprocessor.MyCustomProcessor
+
+        """
+
+    force = Bool(
+        default_value=False,
+        help="Force overwrite of existing files."
+    ).tag(config=True)
+
+    strict = Bool(
+        default_value=False,
+        help="Skip submitted notebooks with invalid names."
+    ).tag(config=True)
+
+    auto_update_database = Bool(
+        default_value=False,
+        help="Automatically update the database and add missing students."
+    ).tag(config=True)
+
+    collect_directory_structure = Unicode(
+        os.path.join(
+            "{downloaded}", "{assignment_id}", "{collect_step}"),
+        help=dedent(
+            """
+            Format string for the directory structure that nbgrader works over
+            during the zip collect process. This MUST contain named keys for
+            'downloaded', 'assignment_id', and 'collect_step'.
+            """
+        )
+    ).tag(config=True)
+
+    downloaded_directory = Unicode(
+        'downloaded',
+        help=dedent(
+            """
+            The main directory that corresponds to the `downloaded` variable in
+            the `collect_structure` config option.
+            """
+        )
+    ).tag(config=True)
+
+    archive_directory = Unicode(
+        'archive',
+        help=dedent(
+            """
+            The name of the directory that contains assignment submission files
+            and/or archives (zip) files manually downloaded from a LMS. This
+            corresponds to the `collect_step` variable in the
+            `collect_structure` config option.
+            """
+        )
+    ).tag(config=True)
+
+    extracted_directory = Unicode(
+        'extracted',
+        help=dedent(
+            """
+            The name of the directory that contains assignment submission files
+            extracted or copied from the `archive_directory`. This corresponds
+            to the `collect_step` variable in the `collect_structure` config
+            option.
+            """
+        )
+    ).tag(config=True)
+
+    zip_ext = List(
+        ['.zip', '.gz', '.tar.gz'],
+        help=dedent(
+            """
+            List of valid archive (zip) filename extensions to extract. Any
+            archive (zip) files with an extension not in this list are copied
+            to the `extracted_directory`.
+            """
+        )
+    ).tag(config=True)
+
+    timezone = Unicode(
+        "UTC",
+        help="Timezone for recording timestamps"
+    ).tag(config=True)
+
+    timestamp_format = Unicode(
+        "%Y-%m-%d %H:%M:%S %Z",
+        help="Format string for timestamps"
+    ).tag(config=True)
+
+    collector_plugin = Type(
```

Comment:
I think unfortunately we're going to have to add support for another type of plugin, which is an extractor plugin. I think in a lot of cases the filenames in people's submissions won't have their usernames in them -- rather, the username will be at a higher level of the directory structure. For example, for Berkeley, we download a giant zip file of all submissions which when extracted has the directory structure like this:

```
Problem Set 1/LAST, FIRST(123456)/Submission attachment(s)/name_of_submission.zip
```

Of course, I could write a separate script to copy these zip files into the `downloaded` directory and rename them. But even them, the files inside won't have the right files names. Even better would be if there was an extractor plugin that could handle both (1) associating the student names with the zipfiles even though there is a directory hierarchy and (2) rename the files within the archive to have student names.